python/ql/src/semmle/python/frameworks/Flask.qll

```diff
@@ -15,127 +15,84 @@ private import semmle.python.ApiGraphs
-
-      /** Gets a reference to the `make_response` method on an instance of `flask.Flask`. */
-      // HACK: We can't call this predicate `make_response` since shadowing is
-      // completely disallowed in QL. I added an underscore to move things forward for
-      // now :(
-      API::Node make_response_() { result = instance_attr("make_response") }
-
-      /**
-       * Gets a reference to the `response_class` attribute on the `flask.Flask` class or an instance.
-       *
-       * See https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.response_class
-       */
-      API::Node response_class() { result = [classRef(), instance()].getMember("response_class") }
+    module View {
+      /** Gets a reference to the `flask.views.View` class or any subclass. */
+      API::Node subclassRef() {
+        result =
+          API::moduleImport("flask")
+              .getMember("views")
+              .getMember(["View", "MethodView"])
+              .getASubclass*()
+      }
     }
 
     /**
-     * Provides models for the `flask.Blueprint` class
+     * Provides models for the `flask.views.MethodView` class and subclasses.
      *
-     * See https://flask.palletsprojects.com/en/1.1.x/api/#flask.Blueprint.
+     * See https://flask.palletsprojects.com/en/1.1.x/views/#method-based-dispatching.
      */
-    module Blueprint {
-      /** Gets a reference to the `flask.Blueprint` class. */
-      API::Node classRef() { result = flask().getMember("Blueprint") }
-
-      /** Gets a reference to an instance of `flask.Blueprint`. */
-      API::Node instance() { result = classRef().getReturn() }
-
-      /**
-       * Gets a reference to the attribute `attr_name` of an instance of `flask.Blueprint`.
-       */
-      private API::Node instance_attr(string attr_name) { result = instance().getMember(attr_name) }
+    module MethodView {
+      /** Gets a reference to the `flask.views.MethodView` class or any subclass. */
+      API::Node subclassRef() {
+        result =
+          API::moduleImport("flask").getMember("views").getMember("MethodView").getASubclass*()
+      }
+    }
+  }
 
-      /** Gets a reference to the `route` method on an instance of `flask.Blueprint`. */
-      API::Node route() { result = instance_attr("route") }
+  /**
+   * Provides models for flask applications (instances of the `flask.Flask` class).
+   *
+   * See https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.
+   */
+  module FlaskApp {
+    /** Gets a reference to the `flask.Flask` class. */
+    API::Node classRef() { result = API::moduleImport("flask").getMember("Flask") }
 
-      /** Gets a reference to the `add_url_rule` method on an instance of `flask.Blueprint`. */
-      API::Node add_url_rule() { result = instance_attr("add_url_rule") }
-    }
+    /** Gets a reference to an instance of `flask.Flask` (a flask application). */
+    API::Node instance() { result = classRef().getReturn() }
+  }
```

Comment:
I feel like this could be encapsulated in a common pattern, say as follows:
```ql
// This should probably live in the API module
class Class extends Node {

  ... // Not really sure about the exact format of the charpred...

  /** Gets an instance of this class. */
  Node getInstance() { result = this.getReturn() }
}

// And then, in the model

class FlaskApp extends API::Class {
  FlaskApp() {this = API::moduleImport("flask").getMember("Flask") }
}
// And then, when used:
predicate foo() {
  blah = any(FlaskApp fa) // for the class
  and
  blah = any(FlaskApp fa).getInstance() // for instances
}
```
With this, we could then also capture the `[FlaskApp::classRef(), FlaskApp::instance()]` pattern as a method of `API::Class` (though I don't know exactly what would be a good name for it).

Reply:
I think being able to capture class-attributes with something more elegant than `[FlaskApp::classRef(), FlaskApp::instance()].getMember("class_attribute_name")` would be nice long term.

However, the fact that `instance()` have a `API::Node` result instead of `DataFlow::Node` goes against my argument in https://github.com/github/codeql/pull/5208/commits/35876f1939dd1ad38ef84870b8fd2063cc80813a :grimacing: (but it really needs `API::Node` to be useful)

So I guess I would like to hold off with this until we're able to use it everywhere. When we have the "def" side of API graphs, it could allow us back in the very familiar QL modeling territory.

Allowing myself to experiment with this idea a bit, with this setup, we could model the "tricky" response class as

```codeql
class FlaskResponse extends API::Class {
  FlaskResponse() {
    this = API::moduleImport("flask").getMember("Response")
    or
    this = any(FlaskApp).getClassMember("response_class")
  }

  override API::Node getInstance() { 
    result = super.getInstance()
    or
    result = API::moduleImport("flask").getMember("make_response").getACall()
    or
    ...
  }
```

However, for some classes, I think we want to model instances as an open union. 

If for example `my_flask_library` also contains functions that can create instances `flask.Response`, with the setup above, we would need to edit the `FlaskResponseClass` implementation in `Flask.qll`, and I would rather not force us into always doing this.

So we could make `FlaskResponseClass` abstract, but then what do we specify in the char-pred for the instance that comes from `my_flask_library.make_response`?

This leads me to think we need a separate modeling of classes and class instances. Possibly something like

```codeql
class FlaskResponseClass extends API::Class {
  FlaskResponseClass() {
    this = "flask.Response"
  }

  override API::Node getARef() {
    this = API::moduleImport("flask").getMember("Response")
    or
    this = any(FlaskApp).getClassMember("response_class")
  }
}

abstract class FlaskResponseInstance extends API::ClassInstance {
  override API::Class getClass() { result instanceof FlaskResponseClass }
}

class FlaskResponseDirectRef extends FlaskResponseInstance {
  FlaskResponseDirectRef() { this = any(FlaskResponseClass cls).getARef().getReturn() }
}

class FlaskMakeResponseCall extends FlaskResponseInstance {
  FlaskMakeResponseCall() { this = API::moduleImport("flask").getMember("make_response").getACall() }
}
```

But again, the modeling of references to a class is not quite open enough, since there is no extension-point to allow `my_flask_library.Response` to be a reference to that class.

**In conclusion**: Great idea, from my point of view, this seems tricky to get right. I would rather have us think this through a bit more, and not implement anything right now.